core/arch/x86/decode.c

```diff
@@ -701,6 +701,119 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
     return pc;
 }
 
+/* Given the potential first evex byte at pc, reads any subsequent evex
+ * bytes (and any prefix bytes) and sets the appropriate prefix flags in di.
+ * Sets info to the entry for the first opcode byte, and pc to point past
+ * the first opcode byte.
+ */
+static byte *
+read_evex(byte *pc, decode_info_t *di, byte instr_byte,
+          const instr_info_t **ret_info INOUT, bool *is_evex)
+{
+
+    const instr_info_t *info;
+    byte prefix_byte = 0, evex_pp = 0;
+    ASSERT(ret_info != NULL && *ret_info != NULL && is_evex != NULL);
+    info = *ret_info;
+
+    CLIENT_ASSERT(info->type == EVEX_PREFIX_EXT, "internal evex decoding error");
+    /* If 32-bit mode and mod selects for memory, this is not vex */
+    if (X64_MODE(di) || TESTALL(MODRM_BYTE(3, 0, 0), *pc)) {
+        /* P[3:2] must be 0 and P[10] must be 1, otherwise #UD */
+        if (TEST(0xC, *pc) || !TEST(0x04, *(pc + 1))) {
+            *ret_info = &invalid_instr;
+            return pc;
+        }
+
+        *is_evex = true;
+        info = &evex_prefix_extensions[0][1];
+    } else {
+        /* not evex */
+        *is_evex = false;
+        *ret_info = &evex_prefix_extensions[0][0];
+        return pc;
+    }
+
+    CLIENT_ASSERT(info->code == PREFIX_EVEX, "internal evex decoding error");
+
+    /* read 2nd evex byte */
+    instr_byte = *pc;
+    prefix_byte = instr_byte;
+    pc++;
+
+    if (TESTANY(PREFIX_REX_ALL | PREFIX_LOCK, di->prefixes) || di->data_prefix ||
+        di->rep_prefix || di->repne_prefix) {
+        /* #UD if combined w/ EVEX prefix */
+        *ret_info = &invalid_instr;
+        return pc;
+    }
+
+    byte evex_mm;
+    CLIENT_ASSERT(info->type == PREFIX, "internal evex decoding error");
+    /* fields are: R, X, B, R', 00, mm.  R, X, B and R' are inverted. */
```

Comment:
Could you mention in a comment that Intel fails to document this in "Vol-2A 2.6 INTEL® AVX-512 ENCODING" but we tested this with common assemblers, as well as the patent WO2012134532A1 mentions that the bits are in fact inverted the same way as in the VEX prefix, in order to make it distinct from the bound instruction in 32-bit mode.